enhancements/api-review/gcp_user_defined_labels.md

```diff
@@ -0,0 +1,610 @@
+
+Motivations include but are not limited to:
+
+- Allow admin, compliance, and security teams to keep track of assets and objects
+  created by OpenShift in GCP.
+
+### User Stories
+
+- As an OpenShift administrator, I want to have labels added to all resources created
+  in GCP by OpenShift, so that I can query resources by the labels or filter resources
+  by label in cloud billing.
+- As an OpenShift administrator, I want to have tags added to all resources created
+  in GCP by OpenShift, so that I can use it for creating IAM policies or to configure 
+  organization policy constraints or to refine resource usage and cost data.
+
+### Goals
+
+- The administrator or service (in the case of Managed OpenShift) installing OpenShift
+  can configure a list of up to 32 user-defined labels in the OpenShift installer generated
+  install config, which is referenced and applied by installer and the in-cluster operators
+  on the GCP resources during creation.
+- Labels must be applied at creation time, in an atomic operation. Labels will not be added post cluster creation.
+- The administrator or service (in the case of Managed OpenShift) installing OpenShift can configure up to 50 tags
+  in the install config. The tags are added by installer and the in-cluster operators while creating GCP resources.
+
+### Non-Goals
+
+- Support for update or delete operation on labels or tags after creation of GCP resource.
+
+## Proposal
+
+***GCP Labels*** 
+
+A label of the form `kubernetes-io-cluster-<cluster_id>:owned` will be added to every
+resource created by OpenShift to enable administrator to differentiate the resources
+created for OpenShift cluster. An administrator is not allowed to add or modify the label
+having the prefix `kubernetes-io` or `openshift-io` in the key. The same can be found
+applied to other cloud platform resources which supports labels/tags for ex: AWS, Azure.
+
+New `userLabels` field will be added to `platform.gcp` of install-config for the user
+to define the labels to be added to the resources created by installer and in-cluster operators.
+
+If `platform.gcp.userLabels` in the install-config has any labels defined, then these labels
+will be added to all the GCP resources created by OpenShift, provided all the defined
+labels meet all the below conditions or else the cluster creation will fail.
+1. A label key and value must have minimum of 1 character and can have maximum of 63 characters.
+2. A label key and value must contain only lowercase letters, numeric characters,
+   underscores, and dashes.
+3. A label key must start with a lowercase letter.
+4. Each resource can have multiple labels, up to a maximum of 32.
+   - GCP supports a maximum of 64 labels per resource. Restricting the number of
+    user defined labels to 32 and reserving 32 for Openshift's internal use.
+
+All in-cluster operators that create GCP resources (Cluster Infrastructure ,Storage) will
+apply these labels during resource creation.
```

Comment:
As for storage, we already do label the volumes (PD: https://github.com/openshift/gcp-pd-csi-driver-operator/pull/16, Filestore: (https://github.com/openshift/gcp-filestore-csi-driver-operator/pull/38). This however requires support also from the CSI drivers, i.e. needs some upstream changes too. I assume it may be similar also for other resource types.

Reply:
Added as limitation in TP.

Reply:
OK. I agree that the labels are fine and we know how to implement that change, so no problem in 4.14 with labels. Tags might be more complicated, so let's postpone that part to 4.15.